### ISSUE TITLE
Create self-signed x509 certificates with Subject Alternative Names

### ISSUE THREAD

**Issue by charlesjohnson** [chef-cookbooks/openssl#37]
It would be very helpful to be able to generate self-signed certificates with SAN fields.

**Comment by charlesjohnson** [comment]
Need to do more reading, but here's the rough outline of how to get this done, I think.

Sources so far:
1) http://stackoverflow.com/questions/21488845/how-can-i-generate-a-self-signed-certificate-with-subjectaltname-using-openssl
2) http://ruby-doc.org/stdlib-2.1.1/libdoc/openssl/rdoc/OpenSSL/Config.html
3) https://mta.openssl.org/pipermail/openssl-users/2015-March/000863.html

First, create a new attribute that would allow a user to submit an openssl config hash to the x509 generator LWRP. Use the config hash as part of the cert generation (but not the private key generation). This is automatically merged into the existing config, which is useful, because if that's true it means our values can be merged into the existing system default config instead of forcing us to generate an entire new openssl config.

The default config for openssl in a chef-client run is at `/opt/chef/embedded/ssl/openssl.cnf`. We can load this as the starting config object to merge on top of with `config = OpenSSL::Config.load(OpenSSL::Config::DEFAULT_CONFIG_FILE)`.

To generate a self-signed cert, it seems the finished openssl config should have these fields:

```
[ alternate_names ]
DNS.1        = example.com
DNS.2        = www.example.com
DNS.3        = mail.example.com
DNS.4        = ftp.example.com

[ v3_ca ]
subjectAltName      = @alternate_names
keyUsage = digitalSignature, keyEncipherment

[ CA_default ]
copy_extensions = copy
```

**Comment by charlesjohnson** [comment]
IP addresses are handled in subjectAltName as `subjectAltName=IP:192.168.7.1`.

so like so:
```
[ alternate_names ]
IP.1 = 192.168.7.1
IP.2 = 127.0.0.1
```
etc.

**Comment by charlesjohnson** [comment]
Paging @jjasghar to this issue.

**Comment by ChefRycar** [comment]
Realized that this change causes some issues using knife-analytics. From the look of it, the above config will cause openssl verify to return a "unable to get local issuer certificate" error rather than the standard "error 18 at 0 depth lookup:self signed certificate" error that it knows what to do with.

tl;dr: Even with the cert in trusted_certs, knife-analytics will throw ssl errors. Example:

```
ubuntu@ip-172-31-18-85:~/update-testing/delivery-demo$ knife rule list
ERROR: SSL Validation failure connecting to host: X.X.X.X - SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed
```

As a fix, I attempted to test adding the following configuration option to keyUsage in the ssl.rb on my testing node's opscode-analytics cookbook:

```keyCertSign```

This DOES generate the correct error, however, it looks like (potentially as a byproduct of of the alt-name) it now thinks there's a hostname mismatch:

```Original Exception: OpenSSL::SSL::SSLError: hostname "X.X.X.X" does not match the server certificate```

I have confirmed that openssl verify reports the appropriate hostname as the CN, which is why my knee-jerk is to think the altname is causing the issue. Have not yet looked at the ssl validation code to see how it's doing hostname matching.

**Comment by charlesjohnson** [comment]
Code snippet that generates an OpenSSL config based on a hash input:

```
require 'openssl'

config = OpenSSL::Config.new

config_hash = {
  'alternate_names' => {
    "DNS.1" => "localhost",
    "IP.1" => "127.0.0.1"
  },
  'v3_ca' => {
    "subjectAltName" => "@alternate_names"
  },
  'CA_default' => {
    "copy_extensions" => "copy"
  }
}

config_hash.keys.each do |section|
  config[section] =  config_hash[section]
end

puts config.to_s
```

**Comment by whitehat101** [comment]
> There's no way to specify subjectAltNames to the x509 certificate generator in the Ruby OpenSSL library.

Actually you can!

From the [OpenSSL::X509::Certificate docs](https://ruby-doc.org/stdlib-2.2.3/libdoc/openssl/rdoc/OpenSSL/X509/Certificate.html), their first example is creating a self signed certificate authority. The trick is that the extensions requested in the CSR don't automatically get copied into the cert's extensions.  In particular, you'd need to copy the SubjectAlternativeName extension over. (Real CA's care a lot about the final cert's Subject and Extensions, blindly copying the extensions could be a security problem, so OpenSSL makes this explicit).

There is a gem, [R509](https://github.com/r509/r509), that provides a high-level abstraction for working with x509. I'm currently using it in a project, but was looking into making a version that only depends on Ruby standard libs.

Adapted example from x509's README:

```ruby
csr = R509::CSR.new(
  subject: [['CN','example.org']],
  san_names: R509::ASN1::GeneralNames.new([{type:'DNS', value:"one.example.org"}, {type:'IP', value:'127.0.0.1'}])
)
ext = []
ext << R509::Cert::Extensions::BasicConstraints.new(:ca => false)
ext << R509::Cert::Extensions::SubjectAlternativeName.new(value: csr.san.to_h) # important
cert = R509::CertificateAuthority::Signer.selfsign(
  csr: csr,
  subject: csr.subject,
  extensions: ext,
)
puts cert.to_pem
```

Because you tickled my interest, I took a second look at the example code from the docs, and wrote this:

```ruby
require 'openssl'
root_key = OpenSSL::PKey::RSA.new 2048 # the CA's public/private key
root_ca = OpenSSL::X509::Certificate.new
root_ca.version = 2 # cf. RFC 5280 - to make it a "v3" certificate
root_ca.serial = 1
root_ca.subject = OpenSSL::X509::Name.parse "/DC=org/DC=ruby-lang/CN=Ruby CA"
root_ca.issuer = root_ca.subject # root CA's are "self-signed"
root_ca.public_key = root_key.public_key
root_ca.not_before = Time.now
root_ca.not_after = root_ca.not_before + 2 * 365 * 24 * 60 * 60 # 2 years validity
ef = OpenSSL::X509::ExtensionFactory.new
ef.subject_certificate = root_ca
ef.issuer_certificate = root_ca
root_ca.add_extension(ef.create_extension("basicConstraints","CA:TRUE",true))
root_ca.add_extension(ef.create_extension("keyUsage","keyCertSign, cRLSign", true))
root_ca.add_extension(ef.create_extension("subjectKeyIdentifier","hash",false))
root_ca.add_extension(ef.create_extension("authorityKeyIdentifier","keyid:always",false))
root_ca.add_extension(ef.create_extension("subjectAltName","DNS:one.example.org,IP:127.0.0.1",false))
root_ca.sign(root_key, OpenSSL::Digest::SHA256.new)
puts root_ca.to_pem
```

Send the output through `openssl x509 -noout -text` and

```
            X509v3 Subject Alternative Name:
                DNS:one.example.org, IP Address:127.0.0.1
```

**Comment by charlesjohnson** [comment]
That is pretty badass. Thanks!

**Comment by tas50** [comment]
Looks like we have this in master now
